Document/0x05c-Reverse-Engineering-and-Tampering.md

```diff
@@ -829,68 +864,343 @@ As of this writing, Android versions up to 4.4.1 run fine in PANDROID, but anyth
+.text:000017EC                 BNE     loc_1854
+.text:000017F0                 LDRB    R4, [R11,#var_1B]
+.text:000017F4                 BL      sub_170C
+.text:000017F8                 MOV     R3, R0
+.text:000017FC                 CMP     R4, R3
+.text:00001800                 BNE     loc_1854
+.text:00001804                 LDRB    R4, [R11,#var_1A]
+.text:00001808                 BL      sub_16F0
+.text:0000180C                 MOV     R3, R0
+.text:00001810                 CMP     R4, R3
+.text:00001814                 BNE     loc_1854
+.text:00001818                 LDRB    R4, [R11,#var_19]
+.text:0000181C                 BL      sub_1728
+.text:00001820                 MOV     R3, R0
+.text:00001824                 CMP     R4, R3
+.text:00001828                 BNE     loc_1854
+.text:0000182C                 LDRB    R4, [R11,#var_18]
+.text:00001830                 BL      sub_1744
+.text:00001834                 MOV     R3, R0
+.text:00001838                 CMP     R4, R3
+.text:0000183C                 BNE     loc_1854
+.text:00001840                 LDR     R3, =(aProductActivat - 0x184C)
+.text:00001844                 ADD     R3, PC, R3      ; "Product activation passed. Congratulati"...
+.text:00001848                 MOV     R0, R3          ; char *
+.text:0000184C                 BL      puts
+.text:00001850                 B       loc_1864
+.text:00001854 ; ---------------------------------------------------------------------------
+.text:00001854
+.text:00001854 loc_1854                                ; CODE XREF: sub_1760+8C
+.text:00001854                                         ; sub_1760+A0 ...
+.text:00001854                 LDR     R3, =(aIncorrectSer_0 - 0x1860)
+.text:00001858                 ADD     R3, PC, R3      ; "Incorrect serial."
+.text:0000185C                 MOV     R0, R3          ; char *
+.text:00001860                 BL      puts
+.text:00001864
+.text:00001864 loc_1864                                ; CODE XREF: sub_1760+F0
+.text:00001864                 SUB     SP, R11, #8
+.text:00001868                 LDMFD   SP!, {R4,R11,PC}
+.text:00001868 ; End of function sub_1760
+```
+
+You can see a loop with some XOR-magic happening at loc_1784, which supposedly decodes the input string. Starting from loc_17DC, you can see a series of decoded values compared with values from further subfunction calls. Even though this doesn't look like highly sophisticated stuff, you'd still need to analyze more to completely reverse this check and generate a license key that passes it. Now comes the twist: dynamic symbolic execution enables you to construct a valid key automatically! The symbolic execution engine maps a path between the first instruction of the license check (0x1760) and the code that prints the "Product activation passed" message (0x1840) to determine the constraints on each byte of the input string. The solver engine then finds an input that satisfies those constraints: the valid license key.
```

Comment:
this might need some references to arm instructions before you continue as this example does not speak for itself as it is described here.

Reply:
Everything in section "##### Symbolic Execution" was already there so I would leave it as it is for this PR and open a ticket addressing concretely this.